pkg/pillar/hypervisor/kvm.go

```diff
@@ -457,23 +457,138 @@ func (ctx kvmContext) Name() string {
+	// 2.5% of the total RAM allocated for VM in bytes
+	// this formula is precise and well explained in the following article:
+	// TBD
+	return int64(config.Memory) * 1024 * 25 / 1000
+}
+
+// overhead for qemu binaries and libraries
+func qemuVMMOverhead() int64 {
+	return 20 << 20 // Mb in bytes
 }
 
-func vmmOverhead(config types.DomainConfig,
-	globalConfig *types.ConfigItemValueMap) (int64, error) {
+// overhead for VMM memory mapped IO
+// it fluctuates between 0.66 and 0.81 % of MMIO total size
+// for all mapped devices. Set it to 1% to be on the safe side
+// this can be a pretty big number for GPUs with very big
+// aperture size (e.g. 64G for NVIDIA A40)
+func mmioVMMOverhead(domainName string, config types.DomainConfig, aa *types.AssignableAdapters) (int64, error) {
+	var pciAssignments []pciDevice
+	var mmioSize uint64
+
+	for _, adapter := range config.IoAdapterList {
+		logrus.Debugf("processing adapter %d %s\n", adapter.Type, adapter.Name)
+		aaList := aa.LookupIoBundleAny(adapter.Name)
+		// We reserved it in handleCreate so nobody could have stolen it
+		if len(aaList) == 0 {
+			return 0, logError("IoBundle disappeared %d %s for %s\n",
+				adapter.Type, adapter.Name, domainName)
+		}
+		for _, ib := range aaList {
+			if ib == nil {
+				continue
+			}
+			if ib.UsedByUUID != config.UUIDandVersion.UUID {
+				return 0, logError("IoBundle not ours %s: %d %s for %s\n",
+					ib.UsedByUUID, adapter.Type, adapter.Name,
+					domainName)
+			}
+			if ib.PciLong != "" && ib.UsbAddr == "" {
+				logrus.Infof("Adding PCI device <%s>\n", ib.PciLong)
+				tap := pciDevice{pciLong: ib.PciLong, ioType: ib.Type}
+				pciAssignments = addNoDuplicatePCI(pciAssignments, tap)
+			}
+		}
+	}
+
+	for _, dev := range pciAssignments {
+		logrus.Infof("PCI device %s %d\n", dev.pciLong, dev.ioType)
+		// read the size of the PCI device aperture
+		if dev.ioType != types.IoOther {
+			continue
+		}
+		// skip bridges
+		isBridge, err := dev.isBridge()
+		if err != nil {
+			// do not treat as fatal error
+			logrus.Warnf("Can't read PCI device class, treat as bridge %s: %v\n",
+				dev.pciLong, err)
+			isBridge = true
+		}
+
+		if isBridge {
+			logrus.Infof("Skipping bridge %s\n", dev.pciLong)
+			continue
+		}
+
+		// read all resources of the PCI device
+		resources, err := dev.readResources()
+		if err != nil {
+			return 0, logError("Can't read PCI device resources %s: %v\n",
+				dev.pciLong, err)
+		}
+
+		// calculate the size of the MMIO region
+		for _, res := range resources {
+			if res.valid() && res.isMem() {
+				mmioSize += res.size()
+			}
+		}
+	}
+
+	mmioOverhead := int64(mmioSize) / 100
+
+	logrus.Infof("MMIO size: %d / overhead: %d for %s", mmioSize, mmioOverhead, domainName)
+
+	return int64(mmioOverhead), nil // 1% of the total MMIO size in bytes
+}
+
+// each vCPU requires about 3MB of memory
+func cpuVMMOverhead(config types.DomainConfig) int64 {
+	return int64(config.MaxCpus) * (3 << 20) // Mb in bytes
```

Comment:
If maxcpus is zero you need to use vcpus; very common to have a zero maxcpus which we elsewhere treat as maxcpus = vcpus e.g.,
  "VCpus": 2,
  "MaxCpus": 0,

Reply:
thanks! Will fix it